### PR TITLE
[Features]mixed precision support enhancement using decorator

### PR DESCRIPTION
**Pull request by Dido0o0**
This feature realize the automatic/constant loss scaling algorithm in Backprop, which can help preserve small gradient magnitudes ( ref to  https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html)
Mixed precision with loss scaling is already support by tf.contrib.mixed_precision.LossScaleOptimizer,  with the following usage:
```python
loss = loss_fn()
opt = tf.AdamOptimizer(learning_rate=...)

# Choose a loss scale manager which decides how to pick the right loss scale
# throughout the training process.
loss_scale_manger = tf.contrib.mixed_precision.FixedLossScaleManager(5000)

# Wraps the original optimizer in a LossScaleOptimizer.
loss_scale_optimizer = LossScaleOptimizer(opt, loss_scale_manager)

# Call minimize() on the loss scale optimizer.
train_op = loss_scale_optimizer.minimize(loss)
```

The optimizer `LossScaleOptimizer` is actually a wrapper of other optimizers.  However, the usage using wrapper is too limited to support all cases.  For example,
    + if the users  are using `tf.gradients()` to compute the gradients rather than `optimizer.compute_gradients()`/`optimizer.minimize()`,  the wrapper is no longer in user. 
    + the `LossScaleOptimizer` wrapper may have compatibility problem with other optimizer wrappers, e.g. `SyncReplicasOptimizer`
    + the current implementation in `LossScaleOptimizer` did not take the colocation parameter into consideration. 

Instead, we implement loss scaling using a decorator on `gradients()`, which is the atom function to compute the backprop gradients. The usage is as following:
```python
loss = loss_fn()
opt = tf.AdamOptimizer(learning_rate=...)
# constant loss scaling with loss_scale=64
with tf.mixed_precision_scope(automatic_loss_scaling=False, loss_scale=64):
      grad = tf.gradients()  # if using tf.gradients
      var_and_grad =opt.minimize(loss) # if using optimizer

# automatic loss scaling
with tf.mixed_precision_scope(automatic_loss_scaling=True):
      grad = tf.gradients()  # if using tf.gradients
      var_and_grad =opt.minimize(loss) # if using optimizer
```

The implementation can handle all of the above problems encounter by `LossScaleOptimizer` .

### CODE REVIEW

**Comment by facaiy** [comment]
@protoget Hi, I think LossScaleOptimizer is introduced by 4c2cb71 . Would you mind taking a look or redirecting to someone else who can? Thank you.

**Comment by wangsiyu** [comment]
This feature has been used internally and simplifies the code change to model function. Please take a look. Thanks. @protoget

**Comment by yangjunpro** [comment]
@protoget Could you please kindly take a look at this CR or delegate it to someone who have the bandwidth? This has already been deployed and successfully used in our in-house environment and we think it should also be beneficial for the community, so we make this PR. 

Thanks.

**Review by protoget — COMMENTED** [review]
Thanks for your hard work and enthusiasm for making valuable contributions to TF community. The code looks well-documented and easy to follow.

One big risk of this approach, is that there's a big difference between "applying zero gradients" and "not applying gradient update" for optimizers that rely on accumulated states. (e.g. Adam). Keep applying zero gradients effectively diluted the acc states, and potentially could cause model quality issues, which we're strongly reluctant to adopt.

To comment on the pain points 
- If the users are using tf.gradients() to compute the gradients rather than optimizer.compute_gradients()/optimizer.minimize(), the wrapper is no longer in user.
Yes. Could you give an example where ```tf.gradients()``` couldn't be replaced by ```optimizer.compute_gradients()```?

- The LossScaleOptimizer wrapper may have compatibility problem with other optimizer wrappers, e.g. SyncReplicasOptimizer
We have to check on that. Most likely we'd fix that single optimizer if that's the case. Do you have an example?

- The current implementation in LossScaleOptimizer did not take the colocation parameter into consideration.
One can use ```optiimizer.compute_gradients()``` and ```optimizer.apply_gradients()``` separately. The first API allows colocation options the same way ```tf.gradients()``` does. ```LossScaleOptimizer``` allows split ```minimize()``` into ```compute_gradients()``` and ```apply_gradients()```.

Also we're a bit concerned about introducing the ```mixed_precision_scope()``` which seems only affect gradient computation. (the name being too broad). In general TF gets complains about how scopes complicate debugging and we try not to add too many scopes.

We appreciate the good work you've accomplished. If the above points do not concern you or your in-house environment, maybe you could commit to tf.contrib (just like ```LossScaleOptimizer``` which is for experimental, non-default features.). But bear in mind tf.contrib is going away in TF2.0. In that case you can have a separate lib and override ```tf.gradients``` to ```tf.alibaba.gradients```.

**Comment by Dido0o0** [comment]
@protoget Thank you for your comments.
As for the issue  "applying zero gradients" versus "not applying gradient update": the case when there is overflow appears is actually in quite low-frequency during the training procedure. In a certain training case, the frequency is approximately 4/10k, which contributes minor impact to the training process. Additionally, we conducted experiments on certain models, e.g. on Faster RCNN with ADAM optimizer, the result is mAP=69.78 (applying zero gradients) versus mAP=69.68 (not applying gradients update). The result is a little bit better. Of course we can say that it is because of random experiments noise, however it can demonstrate that the difference is ignorable.

As for the points you pointed out:
+ If the users are using `tf.gradients()` to compute the gradients rather than `optimizer.compute_gradients()`/`optimizer.minimize()`, the wrapper is no longer in user.
Could you give an example where `tf.gradients()` couldn't be replaced by `optimizer.compute_gradients()`?
This point is mainly for ease of use. Although, the usage of `tf.gradients()` can be replaced by `optimzier.compute_gradients()`, the users need to change more code to take use of `LossScaleOptimzier`. It may introduce more overhead for the users.

+ The LossScaleOptimizer wrapper may have compatibility problem with other optimizer wrappers, e.g. SyncReplicasOptimizer
We have to check on that. Most likely we'd fix that single optimizer if that's the case. Do you have an example?
We have encountered this issue in our in-house model, when using LossScaleOptimizer and SyncReplicasOptimizer. When there is overflow, the apply_gradients will apply no_op, as a result, global_step will not be updated. SyncReplicasOptimizer is synchronized using a token related to global_step. The correlation will trigger hang, when using these two optimizer. Maybe there will be more similar cases for LossScaleOptimizer to be compatible with other optimizers. This is another reason that we choose to apply zeros gradients for overflow case in `tf.gradients()`. The decorator will make `tf.gradients` seems to have totally no difference with/without mixed precision training.

+ The current implementation in LossScaleOptimizer did not take the colocation parameter into consideration.
One can use `optimizer.compute_gradients()` and `optimizer.apply_gradients()` separately. The first API allows colocation options the same way `tf.gradients()` does. LossScaleOptimizer allows split `minimize()` into `compute_gradients()` and `apply_gradients()`.
In loss scale algorithm, we can separate the gradients computation into three parts: 1. scale the loss; 2. backpropagation computation using standard `tf.gradients()` or compute_gradients(); 3. unscale the grads. What you point out that "`optimizer.compute_gradients()`  allows colocation options the same way `tf.gradients()` does" is actually referring the second part. What we would like to point out is the first and third step. The current `LossScaleOptimizer` did not deal with the colocation issue for the 1st and 3rd step. The colocation issue is quite important in certain cases, such as in distribution training.  We have already take it into consideration in the decorator implementation.

Thank you again to review our work in detail with great effort, the above is our consideration to the points you pointed out.

**Comment by protoget** [comment]
Thanks for your response.

* In terms of the parity between 'zero grad' and 'no grad update' -- `tf.gradients()` is a util function and thus should faithfully return the gradient result, and make no assumptions about how client would use the result. How to use the gradient if NaN/Inf, should stay in client code. 
* In terms of the ease of use for `tf.gradients()` vs `opt.compute_gradient()`: maybe, but I don't find it very  concerning. I'm afraid that's not strong enough to justify to zero-out gradients in `tf.gradient()`.
* For "compatibility problem of SyncReplicasOptimzier":
  * (Sorry, but) this optimizer is depreciated.
  * NV suggests just incr the global step even if you get a NaN/Inf grad, just don't update the variables. We did a series of exp and we found that it working.
* For "colocation constraints": I think it's not fundamental limit. The `LossScaleOptimizer' can just do the extra work `gradients_with_scaling()` did.

In summary I think `tf.gradients()` should return gradients as they are, and leave proper handling to client code. We probably couldn't merge this decorator in official TF core, but you can have a library for your production environment, and override the tf implementation when you call it.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @protoget: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by hgadig** [comment]
> Thanks for your response.
> 
> * In terms of the parity between 'zero grad' and 'no grad update' -- `tf.gradients()` is a util function and thus should faithfully return the gradient result, and make no assumptions about how client would use the result. How to use the gradient if NaN/Inf, should stay in client code.
> * In terms of the ease of use for `tf.gradients()` vs `opt.compute_gradient()`: maybe, but I don't find it very  concerning. I'm afraid that's not strong enough to justify to zero-out gradients in `tf.gradient()`.
> * For "compatibility problem of SyncReplicasOptimzier":
>   
>   * (Sorry, but) this optimizer is depreciated.
>   * NV suggests just incr the global step even if you get a NaN/Inf grad, just don't update the variables. We did a series of exp and we found that it working.
> * For "colocation constraints": I think it's not fundamental limit. The `LossScaleOptimizer' can just do the extra work `gradients_with_scaling()` did.
> 
> In summary I think `tf.gradients()` should return gradients as they are, and leave proper handling to client code. We probably couldn't merge this decorator in official TF core, but you can have a library for your production environment, and override the tf implementation when you call it.

@Dido0o0  Hi, could you please respond to the above comments.

**Comment by yangjunpro** [comment]
> > Thanks for your response.
> > 
> > * In terms of the parity between 'zero grad' and 'no grad update' -- `tf.gradients()` is a util function and thus should faithfully return the gradient result, and make no assumptions about how client would use the result. How to use the gradient if NaN/Inf, should stay in client code.
> > * In terms of the ease of use for `tf.gradients()` vs `opt.compute_gradient()`: maybe, but I don't find it very  concerning. I'm afraid that's not strong enough to justify to zero-out gradients in `tf.gradient()`.
> > * For "compatibility problem of SyncReplicasOptimzier":
> >   
> >   * (Sorry, but) this optimizer is depreciated.
> >   * NV suggests just incr the global step even if you get a NaN/Inf grad, just don't update the variables. We did a series of exp and we found that it working.
> > * For "colocation constraints": I think it's not fundamental limit. The `LossScaleOptimizer' can just do the extra work `gradients_with_scaling()` did.
> > 
> > In summary I think `tf.gradients()` should return gradients as they are, and leave proper handling to client code. We probably couldn't merge this decorator in official TF core, but you can have a library for your production environment, and override the tf implementation when you call it.
> 
> @Dido0o0 Hi, could you please respond to the above comments.

Hi hgadig,

Mengdi is on vocation due to the traditional Chinese spring festival so her response may be a little bit late.

As to the comments, actually we have had a phone-call discussion with Google colleagues several months ago and provided our explanations. We will also input our comments into this thread to ensure we share the same background.

Thanks.

**Comment by Dido0o0** [comment]
Hi @hgadig 
As @yangjunpro mentioned , regarding to the above comments, we had online discussion with Google colleagues months ago. The discussion is mainly focused on 1) whether `tf.gradients` should return the 'processed' gradients 2) zero gradients to be applied (this PR) or none gradients  to be applied(`LossScaleOptimizer` ), which is better for the accuracy. 

For 1), in our opinion, once the users turn on loss scaling strategy using `with tf.mixed_precision_scope(automatic_loss_scaling=True):`, the users have already noticed that  
 the gradients results will be handled by the  loss scaling manager,  thus the returned zero-gradients are reasonable and acceptable ( actually I guess the mainly concern is that the code is involved in the core gradients logic);
For 2), actually  both applying zero-gradients and no-gradients applied are approximation method when compared with the fp32 training process. We can not judge which is better. We have conducted experiments with this zero-gradients method upon multiple models, including CNN, RNN and WDL, etc, there are no significant difference to the accuracies obtained by fp32 training.

The advantage of this processing logic with respect to `LossScaleOptimizer` is already mentioned in the above replies. The mainly contribution is that this processing logic can handle the atomic operations of the backpropagation `tf.gradients`, the returned gradients seem to be the same as fp32 bp, thus all the optimizers can be used without any modification.  However `LossScaleOptimizer` is a wrapper for the optimizer (modifying the function `apply_gradients`), which may have potential conflict with other optimizer wrappers.
 
Thanks.

**Comment by reedwm** [comment]
Thank you for the PR, and the detailed justification you provided for putting this in tf.gradients. Currently, our plan is to use the `LossScaleOptimizer`, and not to apply loss scaling in `tf.gradients`. Unfortunately this means I am closing this PR. In addition to the points @protoget already stated, additional reasons for not putting this in tf.gradients are:

* We do not want to add another context manager. Context managers tend to be fairly complicated and we'd like to avoid adding another if necessary.
* In TensorFlow 2, the recommended API is tf.keras, so we are focusing on the optimizer API and not tf.gradients.

All issues with `LossScaleOptimizer`, other than that it is not applied with `tf.gradients`, can be resolved. Also note, we are in the process of moving the `LossScaleOptimizer` outside contrib. See [MixedPrecisionLossScaleOptimizer](https://github.com/tensorflow/tensorflow/blob/1518db83d96aac0efe2c85950581001cfd365821/tensorflow/python/training/experimental/loss_scale_optimizer.py#L30)

This decision is not set in stone. In the upcoming months, we will have a mixed precision design review, where we will describe the entire mixed precision API, including the LossScaleOptimizer.
